CardoExecutor/Common/CardoDataFrame.py

```diff
@@ -1,103 +1,62 @@
-			self._rdd = None
-			self._pandas = None
-		return self._dataframe
-
-	@property
-	def rdd(self):
-		# type: () -> RDD
-		if self._rdd is None:
-			self._rdd = self.dataframe.rdd
-			self._dataframe = None
-			self._pandas = None
-		return self._rdd
-
-	@property
-	def pandas(self):
-		# type: () -> pd.DataFrame
-		if self._pandas is None:
-			self._pandas = self.dataframe.toPandas()
-			self._dataframe = None
-			self._rdd = None
-		return self._pandas
-
-	def persist(self, storage_level=StorageLevel.MEMORY_AND_DISK):
-		persistable_object = getattr(self, self.payload_type)
-		if persistable_object is not None:
-			if hasattr(persistable_object, 'persist'):
-				persistable_object.persist(storageLevel=storage_level)
-				self._persisted.append(persistable_object)
-		return self
-
-	def unpersist(self, **kwargs):
-		blocking = kwargs.get("blocking")
-		for persistable in self._persisted:
-			if blocking is not None:
-				persistable.unpersist(blocking)
-			else:
-				persistable.unpersist()
-
-	@dataframe.setter
-	def dataframe(self, value):
-		self._dataframe = value
-		self._pandas = None
-		self._rdd = None
-
-	@rdd.setter
-	def rdd(self, value):
-		self._rdd = value
-		self._dataframe = None
-		self._pandas = None
-
-	@pandas.setter
-	def pandas(self, value):
-		self._pandas = value
-		self._dataframe = None
-		self._rdd = None
-
-	def deepcopy(self):
-		return CardoDataFrame(None, self.table_name, self._dataframe, self._rdd, copy.deepcopy(self._pandas))
+import pandas as pd
+from CardoExecutor.Common.CardoWrapper import get_wrapped_attribute, _INNER
+from CardoExecutor.Common.CardoContext import CardoContext
+
+
+class CardoDataFrame(DataFrame):
+    """
+    Wraps pyspark.sql.DataFrame. Gives name and a standardised casting between pandas, spark and RDD
+
+    custom __getattribute__ means that methods called on the outer CardoDataFrame object create a new inner variable.
+    """
+    def __init__(self, df: DataFrame, name: str = ""):
+        self._inner = df
+        self.name = name
+
+    def __getattribute__(self, item):
+        return get_wrapped_attribute(self, item)
+
+    def to_cardo_pandas(self):
+        return CardoPandasDataFrame(self._inner.toPandas(), name=self.name)
+
+    def to_cardo_rdd(self):
+        return CardoRDD(self._inner.rdd, name=self.name)
+
+
+class CardoPandasDataFrame(pd.DataFrame):
+    """
+    Wraps pandas.DataFrame. Gives name and a standardised casting between pandas, spark and RDD
+
+    custom __getattribute__ means that methods called on the outer CardoPandasDataFrame object
+    create a new inner variable.
+    """
+    def __init__(self, df: pd.DataFrame, name: str = ""):
+        object.__setattr__(self, _INNER, df)
+        object.__setattr__(self, "name", name)
+
+    def __getattribute__(self, item):
+        return get_wrapped_attribute(self, item)
+
+    def to_cardo_dataframe(self):
+        return CardoDataFrame(CardoContext().createDataFrame(self._inner), name=self.name)
```

Comment:
oh 1 more thing, dont u need to get the spark from the cardo_context?
CardoDataFrame(CardoContext().spark.createDataFrame(self._inner), name=self.name)